src/zenml/deployers/serving/service.py

```diff
@@ -0,0 +1,486 @@
+            "last_execution_time": (
+                self.last_execution_time.isoformat()
+                if self.last_execution_time
+                else None
+            ),
+        }
+
+    def is_healthy(self) -> bool:
+        """Check service health."""
+        return self.snapshot is not None
+
+    # Private helper methods
+
+    def _map_outputs(
+        self,
+        run: PipelineRunResponse,
+        runtime_outputs: Optional[Dict[str, Dict[str, Any]]] = None,
+    ) -> Dict[str, Any]:
+        """Map pipeline outputs using centralized runtime processing."""
+        from zenml.deployers.serving import runtime
+
+        if runtime_outputs is None and runtime.is_active():
+            runtime_outputs = runtime.get_outputs()
+
+        max_size_mb = self._get_max_output_size_bytes() // (1024 * 1024)
+        return runtime.process_outputs(
+            runtime_outputs=runtime_outputs,
+            run=run,
+            enforce_size_limits=True,
+            max_output_size_mb=max_size_mb,
+        )
+
+    def _execute_with_orchestrator(
+        self,
+        resolved_params: Dict[str, Any],
+        use_in_memory: Optional[bool] = None,
+    ) -> PipelineRunResponse:
+        """Run the snapshot via the orchestrator and return the concrete run."""
+        client = self._get_client()
+        active_stack: Stack = client.active_stack
+
+        if self._orchestrator is None:
+            raise RuntimeError("Orchestrator not initialized")
+
+        # Create a placeholder run and execute with a known run id
+        assert self.snapshot is not None
+        placeholder_run = run_utils.create_placeholder_run(
+            snapshot=self.snapshot, logs=None
+        )
+
+        # Start serving runtime context with parameters
+        from zenml.deployers.serving import runtime
+
+        runtime.start(
+            request_id=str(uuid4()),
+            snapshot=self.snapshot,
+            parameters=resolved_params,
+            use_in_memory=use_in_memory,
+        )
+
+        captured_outputs: Optional[Dict[str, Dict[str, Any]]] = None
+        try:
+            self._orchestrator.run(
+                snapshot=self.snapshot,
+                stack=active_stack,
+                placeholder_run=placeholder_run,
+            )
+
+            # Capture in-memory outputs before stopping the runtime context
+            try:
+                if runtime.is_active():
+                    captured_outputs = runtime.get_outputs()
+            except ImportError:
+                pass
+        finally:
+            # Always stop serving runtime context
+            runtime.stop()
+
+        # Fetch the concrete run via its id
+        run: PipelineRunResponse = self._get_client().get_pipeline_run(
+            name_id_or_prefix=placeholder_run.id,
+            hydrate=True,
+            include_full_metadata=True,
+        )
+        # Store captured outputs for the caller to use
+        self._last_runtime_outputs = captured_outputs
```

Comment:
There is no need to store those as part of the object, no? The only place they're used is exactly right after this method returns, so we can simply return them in addition to the run I think